crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env ruby
+#
+# Copyright 2011-2013, Dell
+# Copyright 2013-2016, SUSE Linux GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/cookbook_uploader"
+require "chef/knife/core/object_loader"
+
+module Crowbar
+  class Chef
+    class Upload
+      class << self
+        def all
+          cookbooks
+          data_bags
+          roles
+        end
+
+        def cookbooks
+          version_constraints_to_update = Hash.new
+          cookbooks_for_upload = Array.new
+
+          cookbooks_repo = ::Chef::CookbookLoader.new(chef_cookbooks_path)
+          cookbooks_repo.each do |cookbook_name, cookbook|
+            cookbooks_for_upload.push(cookbook)
+            version_constraints_to_update[cookbook_name] = cookbook.version
+          end
+
+          return unless cookbooks_for_upload.any?
+
+          cookbooks_for_upload.each do |cookbook|
+            logger.info("Checking cookbook #{cookbook.name}")
+            return false unless check_links(cookbook)
+            return false unless check_dependencies(cookbook)
+          end
+
+          cookbook_names = cookbooks_for_upload.map { |cookbook| cookbook.name }
+
+          logger.info("Uploading cookbooks... #{cookbook_names.join(", ")}")
+          ::Chef::CookbookUploader.new(
+            cookbooks_for_upload,
+            force: false,
+            concurrency: cookbooks_for_upload.count
+          ).upload_cookbooks
+        end
+
+        def data_bags
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::DataBagItem, logger)
+          data_bags = loader.find_all_object_dirs(chef_data_bags_path)
+          data_bags.each do |data_bag|
+            data_bag_items = loader.find_all_objects(chef_data_bags_path.join(data_bag))
+            data_bag_item_paths = normalize_data_bag_item_paths(data_bag_items)
+            data_bag_item_paths.each do |data_bag_item_path|
+              # Workaround for a strange chef behavior
+              relative_path = chef_data_bags_path.relative_path_from(Pathname.new(Dir.pwd))
+              data_bag_item = loader.load_from(relative_path, data_bag, data_bag_item_path)
+              bag = ::Chef::DataBagItem.new
+              bag.data_bag(data_bag)
+              bag.raw_data = data_bag_item
+              logger.info("Uploading data_bag #{data_bag_item_path}")
+              bag.save
```

Comment:
I assume it means you copied code from knife? If yes, the copyright header needs to be reflect that (also not sure what's the license for knife).